File: solutions/csharp/allergies/1/Allergies.cs
using System;
using System.Collections.Specialized;
using System.Linq;

public class Allergies
{
    private readonly BitVector32 _levelBits;
    private readonly Allergen _allergen;

    public Allergies(int mask)
    {
        _levelBits = new(mask);
        _allergen = (Allergen)mask;
    }

    public bool IsAllergicTo(Allergen allergen) => _allergen.HasFlag(allergen);

    public Allergen[] List() => Enum.GetValues<Allergen>().Where(a => _levelBits[(int)a]).ToArray();
}

[Flags]
public enum Allergen
{
    Eggs = 1,
    Peanuts = 2,
    Shellfish = 4,
    Strawberries = 8,
    Tomatoes = 16,
    Chocolate = 32,
    Pollen = 64,
    Cats = 128
}

File: solutions/julia/allergies/1/allergies.jl
allergies = Dict(0 => "eggs", 1 => "peanuts", 2 => "shellfish", 3 => "strawberries", 4 => "tomatoes", 5 => "chocolate", 6 => "pollen", 7 => "cats")

allergic_to(score, allergen) = allergen in (allergy_list(score))
allergy_list(score) = allergy_list(score, Set())

function allergy_list(score, list)
    score == 0 && return list
    score == 1 && return push!(list, allergies[0])
	
    l = trunc(Int, log2(score))
    haskey(allergies, l) && push!(list, allergies[l])
    allergy_list(score - 2^l, list)
end